### PR TITLE
Changed EnsureSecurityGroupIncludeDescriptionRule.should_run_rule to …

### PR DESCRIPTION
**Pull request by Tomerva**
…return true if there are security groups. improved security_group.used_by by checking that ENI's that use this SG are valid and are also in use

### CODE REVIEW

**Review by Imanuel-Mizrahi — APPROVED** [review]

File: cloudrail/knowledge/context/aws/ec2/security_group.py [hunk 1]
```diff
@@ -37,7 +38,8 @@ def __init__(self, security_group_id: str, region: str, account: str,
 
     @property
     def used_by(self) -> Set[AwsResource]:
-        return {resource for resource in self._used_by if not resource.invalidation}
+        return {resource for resource in self._used_by if not resource.is_invalidated and
```
Comment (alonashkenazi) [inline cloudrail/knowledge/context/aws/ec2/security_group.py:41]: please add unit test. thx

Comment (Tomerva) [inline cloudrail/knowledge/context/aws/ec2/security_group.py:41]: @yi2020 i have a hard time deciding what to do here..
security_group.used_by is a field that holds the resources (so far ENI's) that use this security group. this is used for rules that check if a security group was used.

Should this set contain only valid resources or can it contain invalid resources?

Right now im leaning toward allowing invalid resources to be part of the used_by property. This is because when a security group is attached to an ENI, and that ENI is attached to a resource, then it means that the SG is used regardless on if we have all the data of the ENI/resource (which could invalidate them).

WDYT?

So to recap, its either not allowing invalidated resource/ENI to be part of the used_by field (the above implementation)
or allow invalid resource (and the implementation will be `if not isinstance(resource, NetworkInterface) or resource.owner`

Comment (Tomerva) [reply cloudrail/knowledge/context/aws/ec2/security_group.py:41]: "Which in turn, will mean the security group is pointing at something invalid" - no, used_by is excluded from the invalidation process. We dont want to invalidate a security group because its attached to an invalid resource, because it could also be attached to some other resource..

**Comment by yi2020** [comment]
Ah good point. And so, you may run into a situation where a security group
will only point to invalid resources. If you remove those resources from
used_by, then used_by may be empty, even though it shouldn't be.

However, we also guarantee a consistent context. All content in the context
should point to valid resources, right? So it makes sense to remove invalid
resources from used_by. If a security group is used by resources, but all
of them are invalid, we should remove the security group.

File: cloudrail/knowledge/context/aws/ec2/security_group.py (hunk 1 above)
Comment (Tomerva) [reply cloudrail/knowledge/context/aws/ec2/security_group.py:41]: Ok i agree, except for the part where we should remove the security group, because it doesnt actually mean its not valid

**Review by alonashkenazi — APPROVED** [review]

File: cloudrail/knowledge/context/aws/ec2/network_interface.py [hunk 2]
```diff
@@ -54,7 +53,7 @@ def __init__(self,
         self.is_primary: bool = is_primary
         self.availability_zone: str = availability_zone
         self.subnet: 'Subnet' = None
-        self.security_groups: List[SecurityGroup] = []
+        self.security_groups: List['SecurityGroup'] = []
```
Comment (yi2020) [inline cloudrail/knowledge/context/aws/ec2/network_interface.py:56]: @Tomerva huh?
